source/extensions/clusters/dynamic_forward_proxy/cluster.h

```diff
@@ -61,8 +65,38 @@ class Cluster : public Upstream::BaseDynamicClusterImpl,
     Upstream::HostConstSharedPtr peekAnotherHost(Upstream::LoadBalancerContext*) override {
       return nullptr;
     }
+    absl::optional<Upstream::SelectedPoolAndConnection>
+    selectPool(Upstream::LoadBalancerContext* context, Upstream::HostConstSharedPtr host,
+               std::vector<uint8_t>& hash_key) override;
+    OptRef<Envoy::Http::ConnectionPool::ConnectionLifetimeCallbacks> lifetimeCallbacks() override;
+
+    // Envoy::Http::ConnectionPool::ConnectionLifetimeCallbacks
+    void onConnectionOpen(Envoy::Http::ConnectionPool::Instance& pool,
+                          std::vector<uint8_t>& hash_key,
+                          const Network::Connection& connection) override;
+
+    void onConnectionDraining(Envoy::Http::ConnectionPool::Instance& pool,
+                              std::vector<uint8_t>& hash_key,
+                              const Network::Connection& connection) override;
 
   private:
+    struct ConnectionInfo {
+      Envoy::Http::ConnectionPool::Instance* pool_; // Not a ref to allow assignment in remove().
+      const Network::Connection* connection_;       // Not a ref to allow assignment in remove().
+    };
+    struct LookupKey {
+      const std::vector<uint8_t> hash_key_;
+      const Network::Address::Instance& peer_address_;
+      bool operator==(const LookupKey& rhs) const {
+        return std::tie(hash_key_, peer_address_) == std::tie(rhs.hash_key_, rhs.peer_address_);
+      }
+    };
+    struct LookupKeyHash {
+      size_t operator()(const LookupKey& /*hash*/) const { return 42; }
+    };
+
+    absl::flat_hash_map<LookupKey, std::vector<ConnectionInfo>, LookupKeyHash> connection_info_map_;
```

Comment:
Ping on this, curious if we need some safeguards to ensure that the proxy behaves in a sane way?

Reply:
Oh, sorry, I lost track of this comment. Thanks for the ping. Yes, I think we should either prevent them from both being active at once or, if that's not possible, disable coalescing if max_concurrent_streams is set. I'll investigate.

Reply:
Ok, I discussed this with @alyssawilk and investigated the code in more detail. As I understand it, all connections have *some* number of max_concurrent_streams (though that limit might be very large). As such we can't really disable this when max_concurrent_streams is used, since it's always used. Instead, what probably makes the most sense is to describe the interaction and allow envoy users the choice of behaviors. WDYT?

Reply:
Sure, if enforcing sane behavior is tricky then I think documenting the behavior would be sufficient

Reply:
Great. The API has been updated to explain this.